cmd/snap/cmd_interface.go

```diff
@@ -0,0 +1,183 @@
+	if len(args) > 0 {
+		return ErrExtraArgs
+	}
+
+	if x.Positionals.Interface != "" {
+		// Show one interface in detail.
+		name := string(x.Positionals.Interface)
+		ifaces, err := Client().Interfaces(&client.InterfaceQueryOptions{
+			Names: []string{name},
+			Doc:   true,
+			Plugs: true,
+			Slots: true,
+		})
+		if err != nil {
+			return err
+		}
+		if len(ifaces) == 0 {
+			return fmt.Errorf(i18n.G("no such interface"))
+		}
+		x.showOneInterface(ifaces[0])
+	} else {
+		// Show an overview of available interfaces.
+		ifaces, err := Client().Interfaces(&client.InterfaceQueryOptions{
+			Connected: !x.ShowAll,
+		})
+		if err != nil {
+			return err
+		}
+		if len(ifaces) == 0 {
+			if x.ShowAll {
+				return fmt.Errorf(i18n.G("no interfaces found"))
+			}
+			return fmt.Errorf(i18n.G("no interfaces currently connected"))
+		}
+		x.showManyInterfaces(ifaces)
+	}
+	return nil
+}
+
+func (x *cmdInterface) showOneInterface(iface *client.Interface) {
+	w := tabwriter.NewWriter(Stdout, 2, 2, 1, ' ', 0)
+	defer w.Flush()
+
+	fmt.Fprintf(w, "name:\t%s\n", iface.Name)
+	if iface.Summary != "" {
+		fmt.Fprintf(w, "summary:\t%s\n", iface.Summary)
+	}
+	if iface.DocURL != "" {
+		fmt.Fprintf(w, "documentation:\t%s\n", iface.DocURL)
+	}
+	if len(iface.Plugs) > 0 {
+		fmt.Fprintf(w, "plugs:\n")
+		for _, plug := range iface.Plugs {
+			if plug.Name == iface.Name {
+				fmt.Fprintf(w, "  - %s", plug.Snap)
+			} else {
+				fmt.Fprintf(w, `  - "%s:%s"`, plug.Snap, plug.Name)
+			}
+			// Print a colon which will make the snap:plug element a key-value
+			// yaml object so that we can write the label or attributes.
+			if plug.Label != "" || len(plug.Attrs) > 0 && x.ShowAttrs {
+				fmt.Fprintf(w, ":\n")
+			} else {
+				fmt.Fprintf(w, "\n")
+			}
+			if plug.Label != "" {
+				fmt.Fprintf(w, "      label:\t%s\n", plug.Label)
+			}
+			x.showAttrs(w, plug.Attrs, "      ")
+		}
+	}
+	if len(iface.Slots) > 0 {
+		fmt.Fprintf(w, "slots:\n")
+		for _, slot := range iface.Slots {
+			if slot.Name == iface.Name {
+				fmt.Fprintf(w, "  - %s", slot.Snap)
+			} else {
+				fmt.Fprintf(w, `  - "%s:%s"`, slot.Snap, slot.Name)
+			}
+			// Print a colon which will make the snap:slot element a key-value
+			// yaml object so that we can write the label or attributes.
+			if slot.Label != "" || len(slot.Attrs) > 0 && x.ShowAttrs {
+				fmt.Fprintf(w, ":\n")
+			} else {
+				fmt.Fprintf(w, "\n")
+			}
+			if slot.Label != "" {
+				fmt.Fprintf(w, "      label:\t%s\n", slot.Label)
+			}
+			x.showAttrs(w, slot.Attrs, "      ")
+		}
+	}
+}
+
+func (x *cmdInterface) showManyInterfaces(infos []*client.Interface) {
+	w := tabWriter()
+	defer w.Flush()
+	fmt.Fprintln(w, i18n.G("Name\tSummary"))
+	for _, iface := range infos {
+		fmt.Fprintf(w, "%s\t%s\n", iface.Name, iface.Summary)
+	}
+}
+
+func (x *cmdInterface) showAttrs(w io.Writer, attrs map[string]interface{}, indent string) {
+	if len(attrs) == 0 || !x.ShowAttrs {
```

Comment:
It's less confusing and less error prone to have the !x.ShowAttrs happening at the call site instead: to not show attrs, don't call show attrs. The len(attrs) == 0 can stay here, though, as this is just the zero case of the function.